### PR TITLE
fix(variant): generates cascades of variants synchronously to avoid timeout dead locks

### LINKED ISSUE #1780: [ANT-871] Timeout error when updating variant properties or matrices with API
## Description

When the user uses the API to update study variants and the variants have a large number of ancestors, we sometimes get a timeout error that crashes processing.

Updates concern both properties and matrices: for example, we have a study with thermal clusters whose parameters are varied with each variant.

The accumulation of timeouts leads to an overload in the number of connections made to the database, gradually bringing the Antares Web server to a standstill. This eventually requires a restart.

## Screenshots

Database connection issue:

MicrosoftTeams-image (1)

## Additional Information

Related stack trace found in the logs:

```Python
Traceback (most recent call last):
File "/antarest/study/storage/variantstudy/variant_study_service.py", line 972, in
  _safe_generation if not self.exists(metadata) and not self._wait_for_generation(metadata, timeout):
File "/antarest/study/storage/variantstudy/variant_study_service.py", line 966, in
  _wait_for_generation self.task_service.await_task(task_id, timeout)
File "/antarest/core/tasks/service.py", line 289, in await_task
  self.tasks[task_id].result(timeout_sec or DEFAULT_AWAIT_MAX_TIMEOUT)
File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 446, in result
  raise TimeoutError() concurrent.futures._base.TimeoutError
```

### PR DESCRIPTION
**Pull request by laurent-laporte-pro**

### CODE REVIEW

**Comment by ghost** [comment]
### Chore

*  **typing:** simplify typing in pytest fixture ([da13c45](https://github.com/AntaresSimulatorTeam/AntaREST/commit/da13c45878f7904c82f105a15a5bbed6036b5f7e))
*  **variant-study-service:** correct the `timeout` parameter to use a default value ([1e2e617](https://github.com/AntaresSimulatorTeam/AntaREST/commit/1e2e6176ba0943c7c53d31f3fbc8e950c50b6188))
* improve log message readability by adding prominent emojis ([eb87bb8](https://github.com/AntaresSimulatorTeam/AntaREST/commit/eb87bb8782a3eb5fca213a64da24bc64e916ed9d))


### Features

*  **repository:** add a method to update a task status to TIMEOUT in database ([38addea](https://github.com/AntaresSimulatorTeam/AntaREST/commit/38addea8ceb246b77bd49793147bec4ebe6baa8e))
*  **variant-generation:** improve error handling for timeout tasks ([979ac6a](https://github.com/AntaresSimulatorTeam/AntaREST/commit/979ac6a4e5cf1cd6c74ee7b5474c389c8cad510f))
*  **tasks:** add timeout parameter to control task completion wait time ([0cac11c](https://github.com/AntaresSimulatorTeam/AntaREST/commit/0cac11c4233a80655323a7eabfb9cac62c22e1df))
*  **variant-study-repo:** add the `get_ancestor_or_self_ids` to retrieve all ancestor IDs of a variant study ([8262e14](https://github.com/AntaresSimulatorTeam/AntaREST/commit/8262e14897dab55f9dad9bd0a21b289f9f481002))
*  **permission:** add the `assert_permission_on_studies` to check user permission of several studies ([ea69c38](https://github.com/AntaresSimulatorTeam/AntaREST/commit/ea69c38574816999046faf0fda8122c559457470))
*  **db:** add methods to check the variant study snapshot status ([e4f7756](https://github.com/AntaresSimulatorTeam/AntaREST/commit/e4f7756a871efe763bb559ba7524fb258aa5532f))
*  **variant:** add the `SnapshotGenerator` class ([70595c9](https://github.com/AntaresSimulatorTeam/AntaREST/commit/70595c982f7b647815b3efe5a78938235d59c60d))
*  **variant:** use the `SnapshotGenerator` class in the `VariantStudyService` ([be805dc](https://github.com/AntaresSimulatorTeam/AntaREST/commit/be805dc62da46cc49dd2cc52bad2ac75d59d9f89))


### Bug Fixes

*  **task-job-service:** update the task status to TIMEOUT if the processing is too long ([ff357bc](https://github.com/AntaresSimulatorTeam/AntaREST/commit/ff357bc6a862c5409ccf31deb0f478ebbb075388))


### Code Refactoring

*  **variant-study-service:** improve implementation of safe generation ([959b2c0](https://github.com/AntaresSimulatorTeam/AntaREST/commit/959b2c021c3e708f0f0b4d5e77e147ac3a71d7c5))
*  **variant-study-repo:** optionally use a user-defined SqlAlchemy session instead of the global `db.session` and improve the documentation ([cd15eec](https://github.com/AntaresSimulatorTeam/AntaREST/commit/cd15eec62302844aed3e740cb3679f8f92da3846))


### Tests

*  **variant-generation:** add a unit test to check that the variant generation is not blocking the main thread ([0ba6320](https://github.com/AntaresSimulatorTeam/AntaREST/commit/0ba63200d12032adbce7795b9da0ce033f87121b))
*  **hydro:** improves testing of hydraulic allocation variants ([21823ba](https://github.com/AntaresSimulatorTeam/AntaREST/commit/21823bae235a8659aa1d2452df76b751ff4b0b05))
*  **variant:** correct the `test_variant_model` unit test to use fixtures ([340881e](https://github.com/AntaresSimulatorTeam/AntaREST/commit/340881e629bfbdfa05ceb1761f0ebe98279c9044))


### Styles

*  **db:** add type hints in database model classes ([4c9c081](https://github.com/AntaresSimulatorTeam/AntaREST/commit/4c9c081da03bde9ce3c1b202562d485099cb7f35))
* correct return type of the get task endpoint ([7dff31a](https://github.com/AntaresSimulatorTeam/AntaREST/commit/7dff31affa0bb1d2c07ae4336d4e16d55cfcba94))
* correct variable naming ([8d42c1b](https://github.com/AntaresSimulatorTeam/AntaREST/commit/8d42c1bd05b6cc1496f60f47eb11e5c99111c1f1))


### Documentation

*  **cache:** add the documentation of cache constants ([94d3cdb](https://github.com/AntaresSimulatorTeam/AntaREST/commit/94d3cdb15b1fe8e3b12612ba832eda6c07366525))


### Performance Improvements

*  **db:** improved study query performance using owner and groups preloading ([16cf54f](https://github.com/AntaresSimulatorTeam/AntaREST/commit/16cf54ff8ef54ce1c7c735a45566ff9e35f96e53))


### Contributors

<a href="https://github.com/laurent-laporte-pro">laurent-laporte-pro</a>

<details>
<summary>Commit-Lint commands</summary>
<br />

You can trigger Commit-Lint actions by commenting on this PR:
- `@Commit-Lint merge patch` will merge dependabot PR on "patch" versions (X.X.Y - Y change)
- `@Commit-Lint merge minor` will merge dependabot PR on "minor" versions (X.Y.Y - Y change)
- `@Commit-Lint merge major` will merge dependabot PR on "major" versions (Y.Y.Y - Y change)
- `@Commit-Lint merge disable` will desactivate merge dependabot PR
- `@Commit-Lint review` will approve dependabot PR
- `@Commit-Lint stop review` will stop approve dependabot PR



</details>

File: antarest/study/repository.py [hunk 1]
```diff
@@ -17,8 +17,30 @@ class StudyMetadataRepository:
...
+        Get the SQLAlchemy session for the repository.
+
+        Returns:
+            SQLAlchemy session.
+        """
+        if self._session is None:
+            # Get or create the session from a context variable (thread local variable)
+            return db.session
+        # Get the user-defined session
+        return self._session
```
Comment (hdinia) [inline antarest/study/repository.py:43]: consider handling exceptions 
```python
 if not self._session:
            try:
                self._session = db.session
            except Exception as e:
                # Raise custom exception

        return self._session
```

File: antarest/core/tasks/repository.py [hunk 2]
```diff
@@ -82,3 +82,11 @@ def delete(self, tid: str) -> None:
...
             db.session.delete(task)
             db.session.commit()
+
+    def update_timeout(self, task_id: str, timeout: int) -> None:
+        """Update task status to TIMEOUT."""
+        task: TaskJob = db.session.get(TaskJob, task_id)
+        task.status = TaskStatus.TIMEOUT
+        task.result_msg = f"Task '{task_id}' timeout after {timeout} seconds"
+        task.result_status = False
+        db.session.commit()
```
Comment (hdinia) [inline antarest/core/tasks/repository.py:92]: here you can also add error handling especially around database operations, and update the task directly if you dont need the entire object for other purposes 
(you can also validate the task_id before the DB operation)

```python
 def update_timeout(self, task_id: str, timeout: int) -> None:
        """Update task status to TIMEOUT."""
        try:
            updated = db.session.query(TaskJob)\
                               .filter(TaskJob.id == task_id)\
                               .update({
                                   'status': TaskStatus.TIMEOUT,
                                    result_msg': f"Task '{task_id}' timeout after {timeout} seconds",
                                   'result_status': False
                               })
            
            if updated:
                db.session.commit()
                logging.info(f"Task {task_id} updated to TIMEOUT status")
            else:
                logging.warning(f"No task found with ID {task_id} to update")

        except SQLAlchemyError as e:
            db.session.rollback()
            logging.error(f"Error updating task {task_id} to TIMEOUT: {e}")
```

File: antarest/core/tasks/web.py [hunk 3]
```diff
@@ -36,17 +38,46 @@ def list_tasks(
...
 
-    @bp.get("/tasks/{task_id}", tags=[APITag.tasks])
+    @bp.get("/tasks/{task_id}", tags=[APITag.tasks], response_model=TaskDTO)
     def get_task(
         task_id: str,
         wait_for_completion: bool = False,
         with_logs: bool = False,
+        timeout: int = DEFAULT_AWAIT_MAX_TIMEOUT,
         current_user: JWTUser = Depends(auth.get_current_user),
     ) -> Any:
```
Comment (hdinia) [inline antarest/core/tasks/web.py]: you can use `TaskDTO` as return type

File: antarest/core/tasks/service.py [hunk 4]
```diff
@@ -283,22 +284,31 @@ def list_db_tasks(self, task_filter: TaskListFilter, request_params: RequestPara
...
-                        break
+                    if task is None:
+                        logger.error(f"Awaited task '{task_id}' was not found")
+                        return
                     if TaskStatus(task.status).is_final():
-                        break
-                    time.sleep(2)
+                        return
+                logger.info("💤 Sleeping 2 seconds...")
+                time.sleep(2)
```
Comment (hdinia) [inline antarest/core/tasks/service.py:308]: if 2s polling interval is not mandatory, consider increasing it, frequent database hits can be inefficient

File: antarest/core/tasks/service.py [hunk 5]
```diff
@@ -283,22 +284,31 @@ def list_db_tasks(self, task_filter: TaskListFilter, request_params: RequestPara
...
+            logger.error(f"Timeout while awaiting task '{task_id}'")
+            with db():
+                self.repo.update_timeout(task_id, timeout_sec)
```
Comment (hdinia) [inline antarest/core/tasks/service.py:311]: what if multiple workers are trying to update the same task simultaneously ?

File: antarest/study/repository.py [hunk 6]
```diff
@@ -44,34 +66,59 @@ def save(
...
-    def get_additional_data(self, study_id: str) -> Optional[StudyAdditionalData]:
+    def get_additional_data(self, study_id: str) -> t.Optional[StudyAdditionalData]:
         metadata: StudyAdditionalData = db.session.query(StudyAdditionalData).get(study_id)
         return metadata
 
-    def get_all(self) -> List[Study]:
+    def get_all(self) -> t.List[Study]:
         entity = with_polymorphic(Study, "*")
-        metadatas: List[Study] = db.session.query(entity).filter(RawStudy.missing.is_(None)).all()
+        metadatas: t.List[Study] = db.session.query(entity).filter(RawStudy.missing.is_(None)).all()
```
Comment (hdinia) [inline antarest/study/repository.py]: `metadatas` as singular: `metadata`
also consider using consistent naming, here `study` `studies` and `metadata` are used, `metadata` looks good in my opinion 
> you can also update it in `get_all_raw` `get_list` etc

File: antarest/study/repository.py [hunk 7]
```diff
@@ -44,34 +66,59 @@ def save(
...
         return studies
 
-    def get_additional_data(self, study_id: str) -> Optional[StudyAdditionalData]:
+    def get_additional_data(self, study_id: str) -> t.Optional[StudyAdditionalData]:
         metadata: StudyAdditionalData = db.session.query(StudyAdditionalData).get(study_id)
         return metadata
 
-    def get_all(self) -> List[Study]:
+    def get_all(self) -> t.List[Study]:
         entity = with_polymorphic(Study, "*")
```
Comment (hdinia) [inline antarest/study/repository.py:113]: here using the polymorphic behavior, I think you will get tuples in the result set

**Review by hdinia — DISMISSED** [review]

**Review by hdinia — APPROVED** [review]

**Pull request merged**